Data race in ipam.TestAllocatorClaim

In a doc PR, so unrelated to changes https://circleci.com/gh/weaveworks/weave/5345
```
git submodule update --init
sudo -E docker run  -ti \
	    -v /home/ubuntu/src/github.com/weaveworks/weave:/go/src/github.com/weaveworks/weave \
		-v /home/ubuntu/src/github.com/weaveworks/weave/.pkg:/go/pkg \
		-e GOARCH -e GOOS -e CIRCLECI -e CIRCLE_BUILD_NUM -e CIRCLE_NODE_TOTAL -e CIRCLE_NODE_INDEX -e COVERDIR -e SLOW \
		weaveworks/weavebuild COVERAGE= WEAVE_VERSION=git-324c1c00813a tests
WARNING: Your kernel does not support memory swappiness capabilities, memory swappiness discarded.
make: Entering directory '/go/src/github.com/weaveworks/weave'

./tools/test -no-go-get

./ipam ./ipam/paxos ./ipam/space

INFO: 2016/04/12 10:34:45.270648 [allocator 01:00:00:01:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.274418 [allocator 01:00:00:01:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.274704 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.386777 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.387342 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.390933 [allocator 01:00:00:02:00:00] Claim 10.0.3.2/24 for b01df00d: address allocator still initializing; will try later.
INFO: 2016/04/12 10:34:45.424220 [allocator 01:00:00:02:00:00] Address 10.0.4.2/24 claimed by abcdef - not in our range
==================

WARNING: DATA RACE

Read by goroutine 16:
  runtime.mapaccess1_fast64()
      /tmp/workdir/go/src/runtime/hashmap_fast.go:95 +0x0
  github.com/weaveworks/weave/testing/gossip.TestRouterClient.GossipUnicast()
      /go/src/github.com/weaveworks/weave/testing/gossip/mocks.go:169 +0x81
  github.com/weaveworks/weave/testing/gossip.(*TestRouterClient).GossipUnicast()
      <autogenerated>:1 +0xf6
  github.com/weaveworks/weave/ipam.(*Allocator).sendRingUpdate()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:868 +0x27d
  github.com/weaveworks/weave/ipam.(*Allocator).donateSpace()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:966 +0x501
  github.com/weaveworks/weave/ipam.(*Allocator).OnGossipUnicast.func1()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:597 +0x1ec
  github.com/weaveworks/weave/ipam.(*Allocator).actorLoop()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:729 +0x19d

Previous write by goroutine 14:
  runtime.mapdelete()
      /tmp/workdir/go/src/runtime/hashmap.go:511 +0x0
  github.com/weaveworks/weave/testing/gossip.(*TestRouter).Stop()
      /go/src/github.com/weaveworks/weave/testing/gossip/mocks.go:45 +0x1ce
  github.com/weaveworks/weave/ipam.stopNetworkOfAllocators()
      /go/src/github.com/weaveworks/weave/ipam/testutils_test.go:238 +0x32
  github.com/weaveworks/weave/ipam.TestAllocatorClaim()
      /go/src/github.com/weaveworks/weave/ipam/allocator_test.go:200 +0xadd
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

Goroutine 16 (running) created at:
  github.com/weaveworks/weave/ipam.(*Allocator).Start()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:123 +0x19a
  github.com/weaveworks/weave/ipam.makeNetworkOfAllocators()
      /go/src/github.com/weaveworks/weave/ipam/testutils_test.go:226 +0x3cb
  github.com/weaveworks/weave/ipam.TestAllocatorClaim()
      /go/src/github.com/weaveworks/weave/ipam/allocator_test.go:167 +0x54
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

Goroutine 14 (running) created at:
  testing.RunTests()
      /tmp/workdir/go/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /tmp/workdir/go/src/testing/testing.go:494 +0xe4
  main.main()
      github.com/weaveworks/weave/ipam/_test/_testmain.go:188 +0x38d
==================
ERRO: 2016/04/12 10:34:45.430856 Dropping message
ERRO: 2016/04/12 10:34:45.431608 Dropping message
ERRO: 2016/04/12 10:34:45.432195 Dropping message
INFO: 2016/04/12 10:34:45.433074 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.433278 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.544793 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.646264 [allocator 01:00:00:01:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.646760 [allocator 01:00:00:01:00:00] Shutdown
INFO: 2016/04/12 10:34:45.647600 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.648127 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.648626 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.699653 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.700070 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.716420 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.716889 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.717393 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.717918 [allocator 03:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:34:45.718344 [allocator 04:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:35:06.601588 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:35:06.606747 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/04/12 10:35:06.614587 [allocator 08:00:27:01:c3:9a] Initialising via deferred consensus
INFO: 2016/04/12 10:35:06.628121 [allocator 08:00:27:01:c3:9a] Initialising via deferred consensus
INFO: 2016/04/12 10:35:06.638211 [allocator 08:00:27:01:c3:9a] Initialising via deferred consensus
PASS

coverage: 60.2% of statements in github.com/weaveworks/weave/ipam, github.com/weaveworks/weave/common, github.com/weaveworks/weave/common/docker, github.com/weaveworks/weave/db, github.com/weaveworks/weave/ipam/paxos, github.com/weaveworks/weave/ipam/ring, github.com/weaveworks/weave/ipam/space, github.com/weaveworks/weave/net/address

Found 1 data race(s)

FAIL	github.com/weaveworks/weave/ipam	22.509s
ok  	github.com/weaveworks/weave/ipam/paxos	24.588s	coverage: 88.2% of statements in github.com/weaveworks/weave/ipam/paxos
ok  	github.com/weaveworks/weave/ipam/space	1.019s	coverage: 35.3% of statements in github.com/weaveworks/weave/ipam/space, github.com/weaveworks/weave/common, github.com/weaveworks/weave/net/address

Makefile:119: recipe for target 'tests' failed
make: *** [tests] Error 1
make: Leaving directory '/go/src/github.com/weaveworks/weave'
make: *** [tests] Error 2
cd $SRCDIR; COVERDIR=test/coverage make RM= tests returned exit code 2
```

--- a/testing/gossip/mocks.go
+++ b/testing/gossip/mocks.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"math"
 	"math/rand"
+	"sync"
 	"time"
 
 	"github.com/weaveworks/mesh"
@@ -32,22 +33,34 @@ type flushMessage struct {
 }
 
 type TestRouter struct {
+	sync.Mutex
 	gossipChans map[mesh.PeerName]chan interface{}
 	loss        float32 // 0.0 means no loss
 }
 
 func NewTestRouter(loss float32) *TestRouter {
-	return &TestRouter{make(map[mesh.PeerName]chan interface{}, 100), loss}
+	return &TestRouter{gossipChans: make(map[mesh.PeerName]chan interface{}, 100), loss: loss}
+}
+
+// Copy so we can access outside of a lock
+func (grouter *TestRouter) copyGossipChans() map[mesh.PeerName]chan interface{} {
+	ret := make(map[mesh.PeerName]chan interface{})
+	grouter.Lock()
+	defer grouter.Unlock()
+	for p, c := range grouter.gossipChans {
+		ret[p] = c
+	}
+	return ret
 }
 
 func (grouter *TestRouter) Stop() {
-	for peer := range grouter.gossipChans {
+	for peer := range grouter.copyGossipChans() {
 		grouter.RemovePeer(peer)
 	}
 }
 
 func (grouter *TestRouter) gossipBroadcast(sender mesh.PeerName, update mesh.GossipData) {
-	for _, gossipChan := range grouter.gossipChans {
+	for _, gossipChan := range grouter.copyGossipChans() {
 		select {
 		case gossipChan <- broadcastMessage{sender: sender, data: update}:
 		default: // drop the message if we cannot send it
@@ -57,8 +70,9 @@ func (grouter *TestRouter) gossipBroadcast(sender mesh.PeerName, update mesh.Gos
 }
 
 func (grouter *TestRouter) gossip(sender mesh.PeerName, update mesh.GossipData) error {
-	count := int(math.Log2(float64(len(grouter.gossipChans))))
-	for dest, gossipChan := range grouter.gossipChans {
+	gossipChans := grouter.copyGossipChans()
+	count := int(math.Log2(float64(len(gossipChans))))
+	for dest, gossipChan := range gossipChans {
 		if dest == sender {
 			continue
 		}
@@ -76,7 +90,7 @@ func (grouter *TestRouter) gossip(sender mesh.PeerName, update mesh.GossipData) 
 }
 
 func (grouter *TestRouter) Flush() {
-	for _, gossipChan := range grouter.gossipChans {
+	for _, gossipChan := range grouter.copyGossipChans() {
 		flushChan := make(chan struct{})
 		gossipChan <- flushMessage{flushChan: flushChan}
 		<-flushChan
@@ -84,11 +98,15 @@ func (grouter *TestRouter) Flush() {
 }
 
 func (grouter *TestRouter) RemovePeer(peer mesh.PeerName) {
+	grouter.Lock()
 	gossipChan := grouter.gossipChans[peer]
+	grouter.Unlock()
 	resultChan := make(chan struct{})
 	gossipChan <- exitMessage{exitChan: resultChan}
 	<-resultChan
+	grouter.Lock()
 	delete(grouter.gossipChans, peer)
+	grouter.Unlock()
 }
 
 type TestRouterClient struct {
@@ -160,13 +178,18 @@ func (grouter *TestRouter) Connect(sender mesh.PeerName, gossiper mesh.Gossiper)
 
 	go grouter.run(sender, gossiper, gossipChan)
 
+	grouter.Lock()
 	grouter.gossipChans[sender] = gossipChan
+	grouter.Unlock()
 	return TestRouterClient{grouter, sender}
 }
 
 func (client TestRouterClient) GossipUnicast(dstPeerName mesh.PeerName, buf []byte) error {
+	client.router.Lock()
+	gossipChan := client.router.gossipChans[dstPeerName]
+	client.router.Unlock()
 	select {
-	case client.router.gossipChans[dstPeerName] <- unicastMessage{sender: client.sender, buf: buf}:
+	case gossipChan <- unicastMessage{sender: client.sender, buf: buf}:
 	default: // drop the message if we cannot send it
 		common.Log.Errorf("Dropping message")
 	}

